service/src/test/java/bio/terra/workspace/service/resource/controlled/GcsApiConversionsTest.java

```diff
@@ -0,0 +1,215 @@
+import com.google.cloud.storage.StorageClass;
+import java.time.Duration;
+import java.time.OffsetDateTime;
+import java.time.ZoneOffset;
+import java.util.Optional;
+import org.junit.jupiter.api.Test;
+
+public class GcsApiConversionsTest extends BaseUnitTest {
+
+  @Test
+  public void testToUpdateParameters() {
+    final ApiGcpGcsBucketUpdateParameters updateParameters1 = toUpdateParameters(GCS_BUCKET_INFO_1);
+    assertEquals(
+        ApiGcpGcsBucketDefaultStorageClass.STANDARD, updateParameters1.getDefaultStorageClass());
+    final ApiGcpGcsBucketLifecycle wsmLifecycle = updateParameters1.getLifecycle();
+    assertThat(wsmLifecycle.getRules(), hasSize(2));
+
+    final ApiGcpGcsBucketLifecycleRule rule1 = wsmLifecycle.getRules().get(0);
+    assertEquals(ApiGcpGcsBucketLifecycleRuleActionType.DELETE, rule1.getAction().getType());
+    assertNull(rule1.getAction().getStorageClass());
+
+    final ApiGcpGcsBucketLifecycleRule rule2 = wsmLifecycle.getRules().get(1);
+    assertEquals(
+        ApiGcpGcsBucketLifecycleRuleActionType.SET_STORAGE_CLASS, rule2.getAction().getType());
+    assertEquals(ApiGcpGcsBucketDefaultStorageClass.STANDARD, rule2.getAction().getStorageClass());
+  }
+
+  @Test
+  public void testToBucketInfo() {
+    final String bucketName = uniqueBucketName();
+    final BucketInfo bucketInfo1 =
+        toBucketInfo(bucketName, ControlledResourceFixtures.BUCKET_UPDATE_PARAMETERS_1);
+    assertEquals(bucketName, bucketInfo1.getName());
+    assertEquals(StorageClass.STANDARD, bucketInfo1.getStorageClass());
+    assertThat(bucketInfo1.getLifecycleRules(), hasSize(2));
+    final LifecycleRule gcsRule1 = bucketInfo1.getLifecycleRules().get(0);
+    assertEquals(DeleteLifecycleAction.TYPE, gcsRule1.getAction().getActionType());
+    assertEquals(31, gcsRule1.getCondition().getAge());
+    assertEquals(toGoogleDateTime(OFFSET_DATE_TIME_2), gcsRule1.getCondition().getCreatedBefore());
+    assertTrue(gcsRule1.getCondition().getIsLive());
+    assertThat(gcsRule1.getCondition().getMatchesStorageClass(), hasSize(2));
+  }
+
+  @Test
+  public void testToBucketInfoNullFields() {
+    final BucketInfo bucketInfo2 = toBucketInfo("bucket-name", BUCKET_UPDATE_PARAMETERS_EMPTY);
+    assertNull(bucketInfo2.getStorageClass());
+    assertThat(bucketInfo2.getLifecycleRules(), is(empty()));
+  }
+
+  @Test
+  public void testLifecycleRuleConversions() {
+    final ApiGcpGcsBucketLifecycleRule wsmRule1 = toWsmApi(GCS_LIFECYCLE_RULE_1);
+    assertEquals(ApiGcpGcsBucketLifecycleRuleActionType.DELETE, wsmRule1.getAction().getType());
+    assertNull(wsmRule1.getAction().getStorageClass());
+
+    assertEquals(42, wsmRule1.getCondition().getAge());
+    assertFalse(wsmRule1.getCondition().isLive());
+    assertEquals(2, wsmRule1.getCondition().getNumNewerVersions());
+    assertNull(wsmRule1.getCondition().getMatchesStorageClass()); // not empty list
```

Comment:
The `BucketInfo` builder converts null to an empty list for lifecycle rules, but the `LifecycleRule` builder does not do that for `matchtesStorageClass`. It's hard to come up with a clean scheme for handling these.

Here are two different conventions for null lists side by side:
```

    /** @deprecated Use {@code setLifecycleRules} method instead. * */
    @Override
    @Deprecated
    public Builder setDeleteRules(Iterable<? extends DeleteRule> rules) {
      this.deleteRules = rules != null ? ImmutableList.copyOf(rules) : null;
      return this;
    }

    @Override
    public Builder setLifecycleRules(Iterable<? extends LifecycleRule> rules) {
      this.lifecycleRules =
          rules != null ? ImmutableList.copyOf(rules) : ImmutableList.<LifecycleRule>of();
      return this;
    }
    ```

Reply:
Well, that is kind of annoying. You have to read everything out and supply everything in, only changing the thing you want to change.